### PR TITLE
Arrow Exporter load prioritizer apparatus

### LINKED ISSUE #147: Add load balancing logic to Arrow exporter
We have setup an Arrow pipeline in which the receiver has (currently) a latency distribution problem. Requests are either long or short, depending on queueing, we've noticed, and the Arrow exporter does nothing to balance load across its streams.

Streams are eager about taking on work. Each time `Send()` finishes for a batch, the stream begins encoding the next batch, regardless of how many items are unacknowledged (which depends on network buffer sizes). 

We hypothesize that some sort of mechanism to prioritize streams with shorter queues will benefit tail latency in our application. An initial strategy we may evaluate:

- let the next unit of work be assigned a priority based on the number of outstanding work items at the moment when the sender is ready for work; new items of work will go to the stream with the shortest queue.

### PR DESCRIPTION
**Pull request by jmacd**
Adds a configurable prioritization scheme to the Arrow exporter, leaving the existing "FIFO" policy the default and adding a new "leastloadedN" policy which considers the least-loaded among N randomly selected streams.

Adds a README.md explaining the internals of the Arrow streaming component. Updates the original diagram.

Adds a benchmark that exercises the different prioritizers and demonstrates different performance characteristics.

This required a substantial redesign of the existing prioritization scheme, which had some unnecessary complexity. The new design:

1. Uses Context cancelation to implement the downgrade signal consistently across prioritizers
2. Simplifies stream-shutdown and restart by re-using the `streamWorkState` object across streams
3. Adds a drain() method to clean up after downgrade, as opposed to the formerly-complex synchronization used.

Extends most tests to cover all prioritizers.

Fixes #147.

### CODE REVIEW

**Review by lquerel — APPROVED** [review]
The documentation and diagram on this significant addition of a new load balancing policy between streams have been very useful. However, I did not see the benchmark results. Were they displayed somewhere?

Thank you for this PR.

File: collector/exporter/otelarrowexporter/config_test.go [hunk 1]
```diff
@@ -86,6 +88,7 @@ func TestUnmarshalConfig(t *testing.T) {
 				MaxStreamLifetime:  2 * time.Hour,
 				PayloadCompression: configcompression.TypeZstd,
 				Zstd:               zstd.DefaultEncoderConfig(),
+				Prioritizer:        "leastloaded8",
```
Comment (moh-osman3) [inline collector/exporter/otelarrowexporter/config_test.go:91]: Should there be a validation error setting N > `num_streams`? i.e. is this something useful for the user to know that they might have changed `num_streams` without updating N for leastloaded prioritizer?

File: collector/exporter/otelarrowexporter/internal/arrow/bestofn.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	}
+}
+
+func (lp *bestOfNPrioritizer) streamFor(_ writeItem, tmp []streamSorter) *streamWorkState {
+	cnt := 0
+	for sws := range lp.state {
+		// TODO: skip channels w/ a pending item (maybe)
+		tmp[cnt].work = sws
+		tmp[cnt].load = lp.loadFunc(sws)
+		if cnt++; cnt == lp.N {
```
Comment (moh-osman3) [inline collector/exporter/otelarrowexporter/internal/arrow/bestofn.go]: Hmm what prevents the last `num_streams - N` items of `lp.state` from being neglected from selection -i.e. it looks like we are always selecting from the first N items of `lp.state`? Is iterating over a map random?

File: collector/exporter/otelarrowexporter/config_test.go (hunk 1 above)
Comment (moh-osman3) [reply collector/exporter/otelarrowexporter/config_test.go:91]: Should N be configured as function of `num_streams`? i.e. as a percentage

File: collector/exporter/otelarrowexporter/internal/arrow/fifo.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+// downgrade indicates that streams are never going to be ready.  Note
+// the caller is required to ensure that setReady() and unsetReady()
+// cannot be called concurrently; this is done by waiting for
+// Stream.writeStream() calls to return before downgrading.
+func (fp *fifoPrioritizer) downgrade(ctx context.Context) {
+	go drain(fp.shared, ctx.Done())
+}
+
+// nextWriter returns the first-available stream.
+func (fp *fifoPrioritizer) nextWriter(ctx context.Context) (streamWriter, error) {
```
Comment (moh-osman3) [inline collector/exporter/otelarrowexporter/internal/arrow/fifo.go]: Will this function ever return an error?

File: collector/exporter/otelarrowexporter/internal/arrow/bestofn.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+		go drain(ws.toWrite, ctx.Done())
+	}
+}
+
+func (lp *bestOfNPrioritizer) sendOne(item writeItem, tmp []streamSorter) {
+	stream := lp.streamFor(item, tmp)
+	writeCh := stream.toWrite
+	select {
+	case writeCh <- item:
+		return
```
Comment (moh-osman3) [inline collector/exporter/otelarrowexporter/internal/arrow/bestofn.go:88]: Do you need to send `item.errCh <- nil` here to signal to waiter before returning?

**Comment by jmacd** [comment]
```
goos: darwin
goarch: arm64
pkg: github.com/open-telemetry/otel-arrow/collector/exporter/otelarrowexporter/internal/arrow
BenchmarkFifo4-10                 	     235	   4996351 ns/op
BenchmarkFifo8-10                 	     198	   6138465 ns/op
BenchmarkFifo16-10                	     152	   7088242 ns/op
BenchmarkFifo32-10                	      79	  13435187 ns/op
BenchmarkFifo64-10                	      48	  20970645 ns/op
BenchmarkFifo128-10               	      90	  12018160 ns/op
BenchmarkLeastLoadedTwo4-10       	     217	   5209409 ns/op
BenchmarkLeastLoadedTwo8-10       	     194	   6073483 ns/op
BenchmarkLeastLoadedTwo16-10      	     138	   8168686 ns/op
BenchmarkLeastLoadedTwo32-10      	      88	  13806171 ns/op
BenchmarkLeastLoadedTwo64-10      	      55	  20729386 ns/op
BenchmarkLeastLoadedTwo128-10     	     162	   7124523 ns/op
BenchmarkLeastLoadedFour4-10      	     198	   5327549 ns/op
BenchmarkLeastLoadedFour8-10      	     183	   6305196 ns/op
BenchmarkLeastLoadedFour16-10     	     134	   8196523 ns/op
BenchmarkLeastLoadedFour32-10     	      58	  17414805 ns/op
BenchmarkLeastLoadedFour64-10     	      74	  18093378 ns/op
BenchmarkLeastLoadedFour128-10    	     142	   7061255 ns/op
PASS
```

TL;DR the first numeric column is number of repetitions to meet the benchmark time threshold (where larger numbers == better), the second column is timing per repetition (smaller == better).  There is a lot of synthetic aspect to this test, many additional channels created by the test apparatus--with a specific and unrealistic latency curve, but it's still meaningful. For the case of 128 streams, FIFO does 90 reps @ 12ms, LeastLoadedTwo does 162 reps @ 7.1ms, LeastLoadedFour does 142 reps @ 7.0ms

File: collector/exporter/otelarrowexporter/internal/arrow/bestofn.go (hunk 2 above)
Comment (jmacd) [reply collector/exporter/otelarrowexporter/internal/arrow/bestofn.go]: Embarassed! I've believed for a long time that Golang map iteration was randomized, but it's in fact just "non-deterministic" and the randomization they apply is not uniform. (Wonder how many of this bug I've written in the last ten years... a couple at least.)

Fixed this. 

```
func (lp *bestOfNPrioritizer) streamFor(_ writeItem, rnd *rand.Rand, tmp []streamSorter) *streamWorkState {
	// Place all streams into the temporary slice.
	for idx, item := range lp.state {
		tmp[idx].work = item
	}
	// Select N at random by shifting the selection into the start
	// of the temporary slice.
	for i := 0; i < lp.N; i++ {
		pick := rand.Intn(lp.N - i)
		tmp[i], tmp[i+pick] = tmp[i+pick], tmp[i]
	}
	for i := 0; i < lp.N; i++ {
		// TODO: skip channels w/ a pending item (maybe)
		tmp[i].load = lp.loadFunc(tmp[i].work)
	}
	sort.Slice(tmp[0:lp.N], func(i, j int) bool {
		return tmp[i].load < tmp[j].load
	})
	return tmp[0].work
}
```

**Comment by jmacd** [comment]
@moh-osman3 Nice catch. The benchmarks improve. :-) 

```
goos: darwin
goarch: arm64
pkg: github.com/open-telemetry/otel-arrow/collector/exporter/otelarrowexporter/internal/arrow
BenchmarkFifo4-10                 	     208	   4954205 ns/op
BenchmarkFifo8-10                 	     201	   5949269 ns/op
BenchmarkFifo16-10                	     152	   7282191 ns/op
BenchmarkFifo32-10                	      68	  15395150 ns/op
BenchmarkFifo64-10                	      50	  22943395 ns/op
BenchmarkFifo128-10               	     105	  11317888 ns/op
BenchmarkLeastLoadedTwo4-10       	     217	   5323038 ns/op
BenchmarkLeastLoadedTwo8-10       	     198	   5880872 ns/op
BenchmarkLeastLoadedTwo16-10      	     153	   6922781 ns/op
BenchmarkLeastLoadedTwo32-10      	     121	   8825361 ns/op
BenchmarkLeastLoadedTwo64-10      	     120	   9001932 ns/op
BenchmarkLeastLoadedTwo128-10     	     309	   3791125 ns/op
BenchmarkLeastLoadedFour4-10      	     192	   5625193 ns/op
BenchmarkLeastLoadedFour8-10      	     170	   6222414 ns/op
BenchmarkLeastLoadedFour16-10     	     146	   7166732 ns/op
BenchmarkLeastLoadedFour32-10     	     128	   8924360 ns/op
BenchmarkLeastLoadedFour64-10     	     122	   9992490 ns/op
BenchmarkLeastLoadedFour128-10    	     306	   3814781 ns/op
```

File: collector/exporter/otelarrowexporter/internal/arrow/fifo.go (hunk 3 above)
Comment (jmacd) [reply collector/exporter/otelarrowexporter/internal/arrow/fifo.go]: Thanks. I've removed the error return, since it isn't used anywhere.

File: collector/exporter/otelarrowexporter/internal/arrow/bestofn.go (hunk 4 above)
Comment (jmacd) [reply collector/exporter/otelarrowexporter/internal/arrow/bestofn.go:88]: In this case, the item of work is moving into the stream for processing, so this is the only branch where item.errCh is not used. Otherwise, all paths lead to `errCh <- ErrStreamRestarting` below.

File: collector/exporter/otelarrowexporter/config_test.go (hunk 1 above)
Comment (jmacd) [reply collector/exporter/otelarrowexporter/config_test.go:91]: I added a `N = min(N, numStreams)` and a comment stating that N would be lowered to match num_streams.

I think it should stay as an integer because the most-useful configuration of this is likely to be a small integer. I do not expect a ratio to be useful, the argument is that there are diminishing returns for raising N such that 2 is really good, 3 is a little better, and no reason to go much further. Here's the blog post on that topic linked in the code: https://smallrye.io/smallrye-stork/1.1.1/load-balancer/power-of-two-choices/

**Comment by jmacd** [comment]
@moh-osman3 Please take another look.

File: collector/exporter/otelarrowexporter/internal/arrow/stream.go [hunk 5]
```diff
@@ -227,22 +229,25 @@ func (s *Stream) run(bgctx context.Context, streamClient StreamClientFunc, grpcO
...
 	}
+
+	s.workState.waiters = map[int64]chan<- error{}
 }
 
 // write repeatedly places this stream into the next-available queue, then
 // performs a blocking send().  This returns when the data is in the write buffer,
 // the caller waiting on its error channel.
-func (s *Stream) write(ctx context.Context) (retErr error) {
+func (s *Stream) write(ctx context.Context, cancel context.CancelFunc) (retErr error) {
```
Comment (moh-osman3) [inline collector/exporter/otelarrowexporter/internal/arrow/stream.go]: Is the `context.CancelFunc` needed in this function?

File: collector/exporter/otelarrowexporter/internal/arrow/exporter.go [hunk 6]
```diff
@@ -127,40 +146,52 @@ func NewExporter(
...
 
-	go e.runStreamController(ctx)
+	// this is the downgradeable context
+	downCtx, downDc := newDoneCancel(ctx)
+
+	var sws []*streamWorkState
+	e.ready, sws = newStreamPrioritizer(downDc, e.prioritizerName, e.numStreams)
+
+	for _, ws := range sws {
+		e.startArrowStream(ctx, ws)
```
Comment (moh-osman3) [inline collector/exporter/otelarrowexporter/internal/arrow/exporter.go]: Thanks for describing the three types of context in the readme. Should this pass in the downCtx instead of the original exporter parent context (like what you do for restarts in runStreamController L194)?

**Review by moh-osman3 — APPROVED** [review]

File: collector/exporter/otelarrowexporter/internal/arrow/stream.go (hunk 5 above)
Comment (jmacd) [reply collector/exporter/otelarrowexporter/internal/arrow/stream.go]: Fixed.

File: collector/exporter/otelarrowexporter/internal/arrow/exporter.go (hunk 6 above)
Comment (jmacd) [reply collector/exporter/otelarrowexporter/internal/arrow/exporter.go]: Thanks-- yes, the README says what I meant.

Comment (jmacd) [reply collector/exporter/otelarrowexporter/internal/arrow/exporter.go]: The reason this isn't a detectable condition now relates to the open TODO about downgrade. The current downgrade logic waits for all of the initial contexts to finish before it can happen.
